### PR TITLE
shredJson() refactor and testing

### PR DESCRIPTION
**Pull request by EricBorczuk**
**What this PR does**:
Refactors shredJson() into more logical pieces; marks `shredForm` as deprecated (it should be deleted after we move away from YCSB testing and it is not documented in the API), and added testing for `shredJson`.

**Which issue(s) this PR fixes**:

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated

### CODE REVIEW

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java [hunk 1]
```diff
@@ -102,97 +120,16 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
...
-                  bindVariableList.add(bindMap.values().toArray());
+                if (value.isValueNode() // scalar or explicit null
+                    || isEmptyObject(value)
+                    || isEmptyArray(value)) {
+                  ImmutablePair<Map<String, Object>, List<String>> result =
+                      convertJsonNodeToBoundVariables(
+                          value, parsingContext, db, path, key, patching);
+                  Map<String, Object> boundVariables = result.left;
+                  firstLevelKeys.addAll(result.right);
+                  logger.debug("{}", boundVariables.values());
```
Comment (tatu-at-datastax) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:131]: I know this is from legacy code but this seems like it's missing any context as to values and perhaps something that really should not be logged at any level? :)
But if it is maybe it should have some info on what those values are ("Bound variables from JSON: {}")

File: restapi/src/test/java/io/stargate/web/docsapi/service/DocsShredderTest.java [hunk 2]
```diff
@@ -0,0 +1,1237 @@
...
+                    surfer, db, Collections.emptyList(), "documentId", payload, false))
+        .hasMessage("Max depth of 64 exceeded.");
+  }
+
+  private List<Object[]> getExpectedRows() {
+    Object[] expectedRow0 =
+        new Object[] {
+          "documentId",
+          "some",
+          "",
```
Comment (tatu-at-datastax) [inline restapi/src/test/java/io/stargate/web/docsapi/service/DocsShredderTest.java:149]: Might be more readable by constructing programmatically, given our Formatter insisting on one element per line...
That is, allocate array, fill with ""s, override ones that are not empty?

File: restapi/src/test/java/io/stargate/web/docsapi/service/DocsShredderTest.java [hunk 3]
```diff
@@ -0,0 +1,1237 @@
...
+          "",
+          "",
+          "structure",
+          null,
+          23.456,
+          null
+        };
+
+    return ImmutableList.of(expectedRow0, expectedRow1, expectedRow2, expectedRow3, expectedRow4);
```
Comment (tatu-at-datastax) [inline restapi/src/test/java/io/stargate/web/docsapi/service/DocsShredderTest.java:1235]: This is fine; JDK has `return Arrays.asList(expectedRow0, ... , expectedRow4);` alternative as well.

**Comment by tatu-at-datastax** [comment]
LGTM; one change that would improve readability is wrt those longish arrays where array declaration is 60+ lines long (due to our Google Formatter's requirements I think?) -- could make less verbose with helper methods.

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java (hunk 1 above)
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:131]: Yeah, upon reading through it again it seems pretty silly to log this - I think waaaaay back in the day we had concerns that this code could be acting wrong and writing unexpected data to C* - but at this point we are pretty confident in its correctness, with testing and all. Removed!

Comment (tatu-at-datastax) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java]: Yeah, it is likely it once served a purpose in figuring out things; nothing wrong with that.

**Comment by EricBorczuk** [comment]
Yes, agree - I was disappointed that there was no way to override the formatter for those lines :P but I will do some gymnastics to make it look less appalling

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java [hunk 4]
```diff
@@ -58,8 +58,12 @@
   public static final String EMPTY_OBJECT_MARKER = "EMPTYOBJ-bccbeee1-6173-4120-8492-7d7bafaefb1f";
   public static final String EMPTY_ARRAY_MARKER = "EMPTYARRAY-9df4802a-c135-42d6-8be3-d23d9520a4e7";
 
+  public static final String LEAF_VALUE_NAME = "leaf";
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentDB.java:61]: I already wrote about this in #1235, any chance we can move to the solutions proposed there wrt constants?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java [hunk 5]
```diff
@@ -73,10 +74,27 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
     if (isJson) {
       return shredJson(surfer, db, path, key, trimmed, patching);
     } else {
+      // This is a deprecated path, and is un-documented
+      // TODO: delete this after no testing relies on it
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:78]: can we delete now?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java [hunk 6]
```diff
@@ -208,6 +144,133 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
...
   }
 
+  private ImmutablePair<Map<String, Object>, List<String>> convertJsonNodeToBoundVariables(
+      JsonNode jsonValue,
+      ParsingContext parsingContext,
+      DocumentDB db,
+      List<String> path,
+      String key,
+      boolean patching) {
+    List<String> firstLevelKeys = new ArrayList<>();
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:154]: can you just move this closer to it's usage.. It's so hard to read this code, if you initialize a var at like `154` and first time use it on line `184`..

Would also be great to add small comment to each private method what's doing and small inline comments in the bigger methods, so somebody can follow the implementation.. It's so hard to read..

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java [hunk 7]
```diff
@@ -208,6 +144,133 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
...
+            DocsApiUtils.convertJsonToBracketedPath(parsingContext.getJsonPath()));
+    Map<String, Object> bindMap = db.newBindMap(path);
+
+    bindMap.put("key", key);
+
+    Iterator<PathOperator> it = jsonPath.iterator();
+    String leaf = null;
+    int i = path.size();
+
+    String[] unboundPaths = new String[DocumentDB.MAX_DEPTH];
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:166]: can we read the max depth from the configuration? you have it just few lines below

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java [hunk 8]
```diff
@@ -208,6 +144,133 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
...
+  }
+
+  private Map<String, Object> addAllToBindMap(
+      Map<String, Object> bindMap,
+      String[] unboundPaths,
+      JsonNode jsonValue,
+      String leaf,
+      boolean treatBooleansAsNumeric) {
+    bindMap = addUnboundPaths(bindMap, unboundPaths);
+    bindMap.put(DocumentDB.LEAF_VALUE_NAME, leaf);
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:203]: See comment above and #1235 fir better way to extract constants to one place

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java [hunk 9]
```diff
@@ -208,6 +144,133 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
...
+    return pathValue;
+  }
+
+  /**
+   * Takes a payload that is a URL-encoded form and turns it into data (bound parameter lists) to
+   * write to a collection. Also returns a list of all "top-level" keys in the form, for the
+   * purposes of merging key sets in the event of a PATCH operation. This is only used by certain
+   * performance tests, is not recommended for use, and will be deleted in the near future.
+   *
+   * @param db
```
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:266]: so this only reason this depends on the DocumentDB is:

* newBindMap
* treatBooleansAsNumeric

Can we extract bind map to some constants helper or whatever.. And have `treatBooleansAsNumeric` as parameter?

This way we can remove DocumentDB dependency completely..

**Review by ivansenic — COMMENTED** [review]
I'll leave to @tatu-at-datastax to approve, as he is more into JSON than me.. If you find any of my proposals good, please go for it.. We can also take some of them to some other ticket.. Let me know..

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java (hunk 5 above)
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java:78]: YCSB tests rely on it, which is the only reason why I wanted to hold off

**Comment by EricBorczuk** [comment]
Hey Ivan - I took the issue you referenced above (#1235) and the comment you made about DocumentDB#newBindMap, and I decided to address it in a PR that I've branched off of this one (as the diff is large): https://github.com/stargate/stargate/pull/1270

Please review!

**Comment by EricBorczuk** [comment]
Alright - I think this PR is ready to go. Let me know!

**Review by tatu-at-datastax — APPROVED** [review]

**Pull request merged**